### PR TITLE
Add Preview langversion and change Default

### PR DESCRIPTION
**Pull request by agocke**
Adds a new "Preview" language version for C# that means the latest
supported language version + the preview features available in the
compiler. Also changes the "Default" (unspecified) language version to
mean Preview, but provides a warning when preview features are used.

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/CodeActions/InlineTemporary/InlineTemporaryTests.cs [hunk 1]
```diff
@@ -3756,7 +3756,7 @@ public void M()
 {
     public void M()
     {
-        var s2 = string.Replace($""hello"", ""world"");
+        var s2 = string.Replace((string)$""hello"", ""world"");
```
Comment (agocke) [inline src/EditorFeatures/CSharpTest/CodeActions/InlineTemporary/InlineTemporaryTests.cs]: @dotnet/roslyn-ide Could someone look at this? It seems strange to me that we need to generate a new cast just because we're using C# 7.3+, but this looks like the current IDE behavior. Is this a bug or am I forgetting something?

Comment (jasonmalinowski) [reply src/EditorFeatures/CSharpTest/CodeActions/InlineTemporary/InlineTemporaryTests.cs]: That's definitely a bug in something. Whether it's the IDE or compiler who knows...

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs [hunk 2]
```diff
@@ -1680,6 +1680,7 @@ internal enum ErrorCode
         ERR_IsNullableType = 8650,
         ERR_AsNullableType = 8651,
 
+        WRN_FeatureInPreview = 8652,
         #endregion diagnostics introduced for C# 8.0
```
Comment (jasonmalinowski) [inline src/Compilers/CSharp/Portable/Errors/ErrorCode.cs]: Blank line before?

File: src/Compilers/CSharp/Portable/LanguageVersion.cs [hunk 3]
```diff
@@ -130,9 +125,25 @@ public enum LanguageVersion
...
+
+        /// <summary>
+        /// The latest supported version of the language.
+        /// </summary>
+        Latest = int.MaxValue - 1,
+
+        /// <summary>
+        /// Preview of the next language version.
         /// </summary>
-        Latest = int.MaxValue,
```
Comment (jasonmalinowski) [inline src/Compilers/CSharp/Portable/LanguageVersion.cs]: This is a breaking change, right? If I recall, enum values are burned into an assembly, so code built against Latest will be considerd as "Preview" (which is appropirate perhaps?) but then upgrading to newer packages and rebuilding causes your burned in value to change?

File: src/EditorFeatures/CSharpTest/CodeActions/InlineTemporary/InlineTemporaryTests.cs (hunk 1 above)
Comment (jasonmalinowski) [reply src/EditorFeatures/CSharpTest/CodeActions/InlineTemporary/InlineTemporaryTests.cs]: In any case: if you don't investigate that prior to merge, I'd say you should instead disable the test (rather than update to an incorrect baseline) and file a tracking bug. The IDE generally does _not_ do the "assert the wrong thing for now" pattern.

File: src/EditorFeatures/CSharpTest/Diagnostics/UpgradeProject/UpgradeProjectTests.cs [hunk 4]
```diff
@@ -604,8 +603,8 @@ void A()
     </Project>
 </Workspace>",
                 new[] {
-                    string.Format(CSharpFeaturesResources.Upgrade_this_project_to_csharp_language_version_0, defaultEffectiveVersion),
-                    string.Format(CSharpFeaturesResources.Upgrade_all_csharp_projects_to_language_version_0, defaultEffectiveVersion)
+                    string.Format(CSharpFeaturesResources.Upgrade_this_project_to_csharp_language_version_0, "8.0 *beta*"),
```
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharpTest/Diagnostics/UpgradeProject/UpgradeProjectTests.cs]: Is there no way for us to fetch that string? Otherwise this will break once you drop the beta tag.

File: src/EditorFeatures/CSharpTest/Diagnostics/UpgradeProject/UpgradeProjectTests.cs [hunk 5]
```diff
@@ -466,7 +466,7 @@ void A()
     </Project>
     <Project Language=""C#"" LanguageVersion=""7"">
     </Project>
-    <Project Language=""C#"" LanguageVersion=""Default"">
+    <Project Language=""C#"" LanguageVersion=""800"">
```
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharpTest/Diagnostics/UpgradeProject/UpgradeProjectTests.cs]: 800?

File: src/Workspaces/MSBuildTest/MSBuildWorkspaceTests.cs [hunk 6]
```diff
@@ -1768,7 +1768,7 @@ public async Task TestParseOptions_CSharp_Compatibility_None()
         public async Task TestParseOptions_CSharp_LanguageVersion_Default()
         {
             CreateCSharpFiles();
-            await AssertCSParseOptionsAsync(CS.LanguageVersion.CSharp7, options => options.LanguageVersion);
+            await AssertCSParseOptionsAsync(CS.LanguageVersion.CSharp8, options => options.LanguageVersion);
```
Comment (jasonmalinowski) [inline src/Workspaces/MSBuildTest/MSBuildWorkspaceTests.cs]: This should probably be some sort of "get effective version of LatestMajor or Default" or something. What's the test actually asserting?

File: src/Compilers/CSharp/Portable/LanguageVersion.cs (hunk 3 above)
Comment (agocke) [reply src/Compilers/CSharp/Portable/LanguageVersion.cs]: That's true, but the change to Default is also a breaking change, and I see no clear way to avoid that, since we're explicitly changing the meaning of Default in this version for the command-line compiler.

File: src/Workspaces/MSBuildTest/MSBuildWorkspaceTests.cs (hunk 6 above)
Comment (agocke) [reply src/Workspaces/MSBuildTest/MSBuildWorkspaceTests.cs]: I read it as testing what the default version is. Do you regard the previous behavior of hard-coding 7.0 as a bug?

**Comment by agocke** [comment]
@jasonmalinowski I'm not against your changes, but for the most part I'm just following the existing behavior/conventions in these files (e.g., hardcoded version names, "800" = 8.0). It seems like changing the style of the tests is reasonable, but out of the scope of this change.

File: src/Workspaces/MSBuildTest/MSBuildWorkspaceTests.cs (hunk 6 above)
Comment (jasonmalinowski) [reply src/Workspaces/MSBuildTest/MSBuildWorkspaceTests.cs]: Yes, that was probbaly a bug.

File: src/Compilers/CSharp/Portable/LanguageVersion.cs (hunk 3 above)
Comment (jasonmalinowski) [reply src/Compilers/CSharp/Portable/LanguageVersion.cs]: Sure, the reinterpretation of default is expected. But you're creating additional breaks which don't seem necessary to me and seem more worrisome. Specifically, a tool creating a Compilation that compiled against "Latest" but running against newer assemblies (consider this happens in Visual Studio!) is now running against "Preview" which is absolutely not what it meant. I get it's "nice" to have the ordering be LatestMajor, Latest, Preview for the sentinel values, but it's causing what seems to me to be avoidable breaks.

I'm also not sure if there's analyzer breaks here too, because what happens if an analyzer wrote "GetEffectiveVersion(LanguageVersion.Latest)"? That'll return preview too.

File: src/Workspaces/MSBuildTest/MSBuildWorkspaceTests.cs (hunk 6 above)
Comment (jasonmalinowski) [reply src/Workspaces/MSBuildTest/MSBuildWorkspaceTests.cs]: And yes, I would expect you to fix this, since this would direct fallout of your design change here.

Comment (agocke) [reply src/Workspaces/MSBuildTest/MSBuildWorkspaceTests.cs]: OK, can do.

File: src/Compilers/CSharp/Portable/LanguageVersion.cs (hunk 3 above)
Comment (agocke) [reply src/Compilers/CSharp/Portable/LanguageVersion.cs]: Makes sense. I'll keep MaxValue as Latest and move the others to smaller values.

**Comment by agocke** [comment]
cc @cartermp

**Comment by agocke** [comment]
@jasonmalinowski I've made some changes, can you take a look? "Default" now means 'latest minor'

File: src/Compilers/CSharp/Portable/Errors/LazyMissingNonNullTypesContextDiagnosticInfo.cs [hunk 7]
```diff
@@ -10,49 +11,69 @@ namespace Microsoft.CodeAnalysis.CSharp
...
+            foreach (var rawInfo in rawInfos)
+            {
+                diagnostics.Add(new LazyMissingNonNullTypesContextDiagnosticInfo(type, rawInfo), location);
+            }
+            rawInfos.Free();
+        }
+
+        private static void GetRawDiagnosticInfos(bool isNullableEnabled, CSharpParseOptions options, ArrayBuilder<DiagnosticInfo> infos)
+        {
+            const MessageID featureId = MessageID.IDS_FeatureNullableReferenceTypes;
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Errors/LazyMissingNonNullTypesContextDiagnosticInfo.cs]: Nit: not sure why this needs to be broken out for just a single usage.

File: src/Compilers/CSharp/Portable/Errors/LazyMissingNonNullTypesContextDiagnosticInfo.cs [hunk 8]
```diff
@@ -10,49 +11,69 @@ namespace Microsoft.CodeAnalysis.CSharp
...
+            var rawInfos = ArrayBuilder<DiagnosticInfo>.GetInstance();
+            GetRawDiagnosticInfos(isNullableEnabled, (CSharpParseOptions)location.SourceTree.Options, rawInfos);
+            foreach (var rawInfo in rawInfos)
+            {
+                diagnostics.Add(new LazyMissingNonNullTypesContextDiagnosticInfo(type, rawInfo), location);
+            }
+            rawInfos.Free();
+        }
+
+        private static void GetRawDiagnosticInfos(bool isNullableEnabled, CSharpParseOptions options, ArrayBuilder<DiagnosticInfo> infos)
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Errors/LazyMissingNonNullTypesContextDiagnosticInfo.cs]: >infos [](start = 131, length = 5)

Do we anticipate needing to add more diagnostics here? Seems like it might be better to just return a tuple of the two DiagnosticInfos that can be returned here, rather than going to the effort of allocating an ArrayBuilder and ensuring that it's freed afterwards.

File: src/EditorFeatures/CSharpTest/Diagnostics/GenerateType/GenerateTypeTests.cs [hunk 9]
```diff
@@ -1571,7 +1571,7 @@ internal class T
 {
     public T(out DateTime d)
     {
-        d = default(DateTime);
+        d = default;
```
Comment (333fred) [inline src/EditorFeatures/CSharpTest/Diagnostics/GenerateType/GenerateTypeTests.cs]: Why make these changes? They don't seem related to the overall change here.

**Comment by 333fred** [comment]
Done review pass (commit 7)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 10]
```diff
@@ -49063,25 +49081,39 @@ void symbolValidator(ModuleSymbol m)
...
                 .Concat(new[] {
                 // error CS8630: Invalid 'NullableContextOptions' value: 'Enable' for C# 7.3. Please use language version 8.0 or greater.
                 Diagnostic(ErrorCode.ERR_NullableOptionNotAvailable).WithArguments("NullableContextOptions", "Enable", "7.3", "8.0").WithLocation(1, 1),
                 }).ToArray()
                 );
+
+            comp = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue(), parseOptions: TestOptions.RegularDefault, skipUsesIsNullable: true);
+            comp.VerifyDiagnostics(expected
+                .Concat(new[] {
+                // error CS8630: Invalid 'NullableContextOptions' value: 'Enable' for C# 7.3. Please use language version 8.0 or greater.
```
[the remark this answers is not in the document]
Comment (agocke) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: That makes sense, I'll look into it.

File: src/EditorFeatures/CSharpTest/Diagnostics/GenerateType/GenerateTypeTests.cs (hunk 9 above)
Comment (agocke) [reply src/EditorFeatures/CSharpTest/Diagnostics/GenerateType/GenerateTypeTests.cs]: These tests use the default language version, which used to be 7.0. When the default is changed to 7.3, the IDE takes advantage of the new "target-typed default" feature, which changes the generated code. This is the new expected baseline for this test.

File: src/Compilers/CSharp/Portable/Errors/LazyMissingNonNullTypesContextDiagnosticInfo.cs (hunk 7 above)
Comment (agocke) [reply src/Compilers/CSharp/Portable/Errors/LazyMissingNonNullTypesContextDiagnosticInfo.cs]: I found it more readable

File: src/Compilers/CSharp/Portable/Errors/LazyMissingNonNullTypesContextDiagnosticInfo.cs (hunk 8 above)
Comment (agocke) [reply src/Compilers/CSharp/Portable/Errors/LazyMissingNonNullTypesContextDiagnosticInfo.cs]: This leaves the door open to adding any number of diagnostics, which I think is worth it. The allocations are pooled, so I'm not worried about that.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (agocke) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: It's a good idea, but I think this is fine to keep as-is. Too many baselines to change and not a whole lot of time.

**Comment by agocke** [comment]
@dotnet/roslyn-compiler  Ping for another review

File: src/Compilers/CSharp/Portable/Parser/Lexer.cs [hunk 11]
```diff
@@ -939,28 +939,11 @@ private void ScanSyntaxToken(ref TokenInfo info)
...
-            var options = this.Options;
-            if (options.IsFeatureEnabled(feature))
-            {
-                return;
-            }
-
-            string requiredFeature = feature.RequiredFeature();
-            if (requiredFeature != null)
+            var info = feature.GetFeatureAvailabilityDiagnosticInfo(Options);
+            if (info != null)
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Parser/Lexer.cs]: This PR seems to jump between `info != null` and `!(info is null)`. Is there a reason?

Comment (agocke) [reply src/Compilers/CSharp/Portable/Parser/Lexer.cs]: No, not really. I was probably just editing the file at a different time and forgot what the prevailing style was. I generally like `!(info is null)`.

**Comment by agocke** [comment]
@333fred Any more comments?

**Comment by agocke** [comment]
ping @jasonmalinowski

**Comment by jaredpar** [comment]
Looks like we're missing the change to the build targets files to condition preview on the target framework.

File: src/EditorFeatures/CSharpTest/Diagnostics/UpgradeProject/UpgradeProjectTests.cs [hunk 12]
```diff
@@ -531,18 +531,19 @@ void A()
         [Fact]
         public async Task OnlyOfferFixAllProjectsFromCSharp6ToDefaultWhenApplicable()
         {
+            var defaultVersion = LanguageVersion.Default.MapSpecifiedToEffectiveVersion().ToDisplayString();
```
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharpTest/Diagnostics/UpgradeProject/UpgradeProjectTests.cs]: Nitpick: I'd call this effectiveVersionOfDefaultVersion. Otherwise the asserts below make it look like we're asserting it *does* say "default" which is not the case.

File: src/Features/CSharp/Portable/UpgradeProject/CSharpUpgradeProjectCodeFixProvider.cs [hunk 13]
```diff
@@ -61,6 +52,11 @@ private static LanguageVersion RequiredVersion(ImmutableArray<Diagnostic> diagno
                 {
                     max = max > required ? max : required;
                 }
+                else if (diagnostic.Id == "CS8652")
```
Comment (jasonmalinowski) [inline src/Features/CSharp/Portable/UpgradeProject/CSharpUpgradeProjectCodeFixProvider.cs]: Is there any reason CS8652 doesn't set DiagnosticPropertyConstants.RequiredLanguageVersion?

Comment (jasonmalinowski) [reply src/Features/CSharp/Portable/UpgradeProject/CSharpUpgradeProjectCodeFixProvider.cs]: (I wouldn't block this PR, but still a good question. Not sure what other code might exist that's grabbing that property and expecting uniformity.)

**Comment by tannergooding** [comment]
@agocke, the OP indicates:
> Adds a new "Preview" language version for C#

Does this mean the same keyword will not be recognized by VB? This makes it so that projects mixing both need to use the following, if they want to control this centrally:
```
<LangVersion>latest</LangVersion>
<LangVersion Condition="'$(Language)' == 'C#'">preview</LangVersion>
```

File: src/Features/CSharp/Portable/UpgradeProject/CSharpUpgradeProjectCodeFixProvider.cs (hunk 13 above)
Comment (agocke) [reply src/Features/CSharp/Portable/UpgradeProject/CSharpUpgradeProjectCodeFixProvider.cs]: Special versions like "preview" aren't, as you noted before, well-ordered in the version list. That means that fixers (like the upgrade project fixer) may produce bad results if they're naively looking at that value and doing simple integer comparison.

Opting-in seems safer, since no one could have possibly seen this diagnostic before.

**Comment by agocke** [comment]
@tannergooding Correct

**Comment by tannergooding** [comment]
That's unfortunate. At least it's easy to workaround though 😄

**Comment by agocke** [comment]
@jaredpar I think this is ready to merge whenever you're ready
